### DISCUSSION TITLE
Issue: `$.[?(... || ...)]` Appears Limited in Mockoon 9.0.0

### DISCUSSION THREAD

**Discussion by tabkram** [mockoon/mockoon#1617] [Help] [answered]
#### Description

After upgrading Mockoon from **v8.4** to **v9.0.0**, I noticed an issue with `jsonParse` using JSONPath. Specifically, queries with more than 14 conditions (`||`) in a filter return an **empty response**, while they worked correctly in version **8.4**.

#### Example Query

Here’s a simplified example of the query:

```
$.[?(
    @.label === "Value 1" ||
    @.label === "Value 2" ||
    ...
    @.label === "Value 14" ||
    @.label === "Value 15"
)]
```

In the above, the actual query contains MORE THAN 14 `||` conditions to match specific labels. 

This worked as expected in **8.4**, but in **9.0.0**, the response is empty if the condition count exceeds 14.

#### Context

I noticed a [potential breaking change in Mockoon 9.0.0](https://mockoon.com/releases/9.0.0/), where a fix for a **JSONPath RCE vulnerability** was introduced. 
Could this change be related to the behavior?

I also reviewed [GitHub Discussion #1582](https://github.com/mockoon/mockoon/discussions/1582), which mentions JSONPath issues in version 9.0.0, but their issue seems to involve `&&` and nested conditions. 

My case differs, as it specifically involves the **number of `||` conditions**.

#### Questions

1. **Is there a known limit** to the number of conditions allowed in JSONPath filters (`||`) in Mockoon 9.0.0?
2. If this is intentional or a regression, **are there plans to address it in future updates**?
3. **Are there any recommended workarounds** to handle queries with a large number of conditions?

#### Notes

- **Environment**:
    - Mockoon CLI v9.0.0.
    - Issue is reproducible consistently across all mocks.

Thank you for your help!

**Comment by 255kb** [comment]
There was indeed a change (https://github.com/mockoon/mockoon/issues/1551) where we had to whitelist some syntaxes with a regex to mitigate an RCE vulnerability. 
It should cover most needs and allow long enough filters like yours.

I don't know what is wrong in your setup, but I just tested to get data from a data bucket with a long filter like you describe and it worked.
Here is the example environment with a minimal setup: 
[sandbox.json](https://github.com/user-attachments/files/18020619/sandbox.json)

(Note that JSONPath used in templating helpers or rules are going through the same regex/checks. )

Maybe you could share your setup?

**Comment by tabkram** [comment]
Thank you for your response @255kb .

After further investigation on my case, I’ve discovered that the issue is not related to the number of conditions in the JSONPath but rather to **the total length of the JSONPath expression**.

Specifically, when the length of the expression **exceeds 1000 characters**, the response is no longer returned, regardless of the number of conditions.

Here’s a minimal setup I’ve created to demonstrate the issue. You can find the setup in the following link:  
[test-setup.json](https://github.com/user-attachments/files/18135533/test-setup.json)

### Example of Working Response (Extractor Length = 1000)

When the total length of the JSONPath expression is **1000 characters**, the query works as expected:

```json
{
    "count": 1,
    "extractorLength": 1000,
    "extractor": "$.[?(@.label==='label1' || @.label==='I am a very long text label here to represent a descriptive  value that spans multiple characters for testing purposes and continues to extend without any special characters just to ensure it can handle scenarios where large amounts of plain text are needed for various testing applications across different systems interfaces or environments I am a very long text label here to represent a lengthy and descriptive value that spans multiple characters for testing purposes and continues to extend without any special characters just to ensure it can handle scenarios where large amounts of plain text are needed for various testing applications across different systems interfaces or environments I am a very long text label here to represent a lengthy and descriptive value that spans multiple characters for testing purposes and continues to extend without any special characters just to ensure it can handle.' || @.label==='TOTAL LENGTH MORE THAN 1000 : no') ]",
    "data": [
        {
            "id": 1,
            "label": "label1"
        }
    ]
}
```

### Example of Non-working Response (Extractor Length > 1000)

When the total length exceeds **1000 characters**, the query fails and returns no results:

```json
{
    "count": 0,
    "extractorLength": 1001,
    "extractor": "$.[?(@.label==='label1' || @.label==='I am a very long text label here to represent a descriptive  value that spans multiple characters for testing purposes and continues to extend without any special characters just to ensure it can handle scenarios where large amounts of plain text are needed for various testing applications across different systems interfaces or environments I am a very long text label here to represent a lengthy and descriptive value that spans multiple characters for testing purposes and continues to extend without any special characters just to ensure it can handle scenarios where large amounts of plain text are needed for various testing applications across different systems interfaces or environments I am a very long text label here to represent a lengthy and descriptive value that spans multiple characters for testing purposes and continues to extend without any special characters just to ensure it can handle.' || @.label==='TOTAL LENGTH MORE THAN 1000 : yes') ]",
    "data": []
}
```

### Investigation Findings

From my testing, I can confirm that the issue occurs not due to the **number of conditions** in the filter but because the **total length of the JSONPath expression exceeds 1000 characters**. This seems to be a limitation in Mockoon 9.0.0 that prevents the query from being processed when the expression length surpasses this threshold.

### Question

Is there a change introduced in Mockoon 9.0.0 related to the handling of long JSONPath queries or a character length limitation that could explain this behavior? If so, is there a way to increase this limit or handle queries with longer expressions?

Any assistance or clarification would be greatly appreciated!

**Reply by 255kb** [reply]
Yes, JSONPath expressions are limited to 1000 characters. Do you really need more? 
I think you can shorten your expression using a regex: `$.[?(@property === "label" && @.match(/value [1234]/i))]^`

**Reply by tabkram** [reply]
Thank you for your response @255kb .  

Yes, in my case, the limitation is restrictive. 
Unfortunately, my real-world JSONPath conditions are more complex than the example I provided and cannot be replaced by a simple regex.  

In Mockoon 8.4, there was no such length limit. Was this limitation introduced natively in JSONPath, or is it a restriction added in Mockoon 9.0 for security or other reasons?  

Is there any way to increase or bypass this limit?

**Reply by 255kb** [reply, marked as answer]
yes the limit is part of the security update. And there is no way to bypass this limit for now. 
I'm not sure it should be raised as 1000 characters is already a lot, and considering there are regex alternatives.

**Reply by tabkram** [reply]
Thank you for your response @255kb .
